python/tvm/relay/testing/py_converter.py

```diff
@@ -0,0 +1,616 @@
+PROLOGUE = [
+    ast.Import([alias('numpy', None)]),
+    ast.Import([alias('tvm', None)]),
+    ast.ImportFrom('tvm', [alias('relay', None)], 0),
+    ast.ImportFrom('tvm.relay.backend.interpreter',
+                   [alias('RefValue', None),
+                    alias('TupleValue', None),
+                    alias('TensorValue', None),
+                    alias('ConstructorValue', None)],
+                   0)
+]
+
+class PythonConverter(ExprFunctor):
+    '''Functor for translating Relay programs into Python ASTs.'''
+
+    def __init__(self, mod, target) -> None:
+        super().__init__()
+        self.mod = mod
+        self.tgt = target
+        self.engine = compile_engine.get()
+        self.fun_no = 0
+        self.var_no = 0
+        self.var_map = {}
+
+
+    def convert(self, prog: Expr):
+        '''This method converts the passed Relay expression into a Python
+        AST object with equivalent semantics.
+
+        The Python AST can be executed using exec(); it can be turned
+        into text and inspected using astor.
+        '''
+        optimized = self.optimize(prog)
+
+        # start with conversion prelude (imports) and convert global defs
+        body = []
+        body += PROLOGUE
+        body += self.convert_module()
+
+        prog_body, extra_defs = self.visit(optimized)
+        body += extra_defs
+
+        # we finally must assign the final expression to the output var
+        # so it can be read after running EXEC
+        body.append(Assign([Name(OUTPUT_VAR_NAME, Store())], prog_body))
+
+        return ast.fix_missing_locations(ast.Module(body=body))
+
+
+    def optimize(self, prog: Expr):
+        '''Performs optimizations necessary to be able to generate code for prog.'''
+        # unwrap tuple wrappers (some op calls produce them)
+        unwrapped = prog.astuple() if isinstance(prog, relay.TupleWrapper) else prog
+        check = relay.ir_pass.infer_type(unwrapped, self.mod)
+        assert relay.ir_pass.well_formed(check)
+        # necessary pass: SimplifyInference (otherwise we can't generate code for some operators)
+        # and fusion (to get primitive functions)
+        simplify = relay.ir_pass.simplify_inference(check)
+        simplify_checked = relay.ir_pass.infer_type(simplify, self.mod)
+        fused = relay.ir_pass.fuse_ops(simplify_checked, opt_level=0)
+        fused_checked = relay.ir_pass.infer_type(fused, self.mod)
+        assert relay.ir_pass.well_formed(fused_checked)
+        return fused_checked
+
+
+    def sanitize(self, name: str) -> str:
+        '''Removes any invalid characters (only underscores, numbers, and letters permitted)
+        from the given name. Since we append a number and underscore to var names anyway,
+        it doesn't matter if the name is the empty string.'''
+        return re.sub(r'\W', '', name)
+
+
+    # generates a unique variable name starting from the hint
```

Comment:
Should we keep the doc format consistent, i.e. moving them under `def`?
Same to the functions below.

Reply:
For things that were meant as purely internal functions, I just put the comments on top but I will change it to be consistent, thanks for the response.